src/xb-arena.c

```diff
@@ -0,0 +1,137 @@
+        memcpy(new_chunklist, arena->chunks, arena->n_chunks * sizeof(uint8_t*));
+        free(arena->chunks);
+        arena->chunks = new_chunklist;
+        arena->chunks_cap *= 2;
+    }
+    arena->chunks[arena->n_chunks] = chunk;
+    arena->n_chunks++;
+}
+
+void* xb_arena_alloc(XbArena* arena, size_t len) {
+    if(len == 0) return NULL;
+    
+    /* Treat allocations larger than a chunk size as their own chunk
+     * It's not worth trying to optimize those. If this case hits a lot, try
+     * increasing the arena chunk size. */
+    if(len > XB_ARENA_CHUNKSIZE){
+        uint8_t* chunk = malloc(len);
+        xb_arena_add_chunk(arena, chunk);
+        /* No need to invalidate tail or current_chunk_end, there's still space
+        in the current chunk for more allocations */
+        return chunk;
+    }
+    /* Round allocations up to nearest whole machine word
+     * If we didn't do this, allocated structs may cost extra
+     * to access because of unaligned load/stores and may
+     * straddle multiple cache lines */
+    len = (len+sizeof(ssize_t)-1) & ~(sizeof(ssize_t)-1);
+
+    /* Allocate another chunk if the the requested size won't fit in the current one.
+     * This will always be hit for the first allocation on a new arena
+     * because tail and current_chunk_end are initialized to zero. */
+    if(arena->tail + len > arena->current_chunk_end) {        
+        uint8_t* chunk = malloc(XB_ARENA_CHUNKSIZE);
+        xb_arena_add_chunk(arena, chunk);
+        arena->tail = chunk;
+        arena->current_chunk_end = chunk + (XB_ARENA_CHUNKSIZE);
+    }
+
+    uint8_t* alloc = arena->tail;
+    arena->tail += len;
+    return (void*)alloc;
+}
+
+void xb_arena_free(XbArena* arena) {
+    for (uint32_t i=0; i < arena->n_chunks; i++) {
+        free(arena->chunks[i]);
+    }
+    free((void*)arena->chunks);
+    arena->n_chunks = 0;
+    arena->chunks_cap = 0;
+    arena->tail = NULL;
+    arena->current_chunk_end = NULL;
+}
+
+char* xb_arena_strdup(XbArena* arena, const char* src) {
+    if(!src) return NULL;
+    int n = strlen(src) + 1;
+    char* out = xb_arena_alloc(arena, n);
+    memcpy(out, src, n);
+    out[n] = '\0';
+    return out;
+}
+
+char* xb_arena_strndup(XbArena* arena, const char* src, size_t strsz) {
+    if(!src) return NULL;
+    /* strnlen isn't in the c99/posix standard even though it's widely supported
+     * easy enough to do it ourselves */
+    size_t n;
+    for(n=0; n<strsz; n++) {
+        if(src[n] == '\0') break;
+    }
+    char* out = xb_arena_alloc(arena, n+1);
+    memcpy(out, src, n);
+    out[n] = '\0';
+    return out;
+}
+
+XbArenaPtrArray* xb_arena_ptr_array_new(XbArena* arena) {
+    XbArenaPtrArray* self = xb_arena_alloc(arena, sizeof(XbArenaPtrArray));
+    *self = (XbArenaPtrArray){
+        .arena = arena,
+        .pointers = (void**) xb_arena_alloc(arena, 4 * sizeof(void*)),
+        .len = 0,
+        .cap = 4,
```

Comment:
Was this cap chosen based on some testing? It would be quite a penalty to reallocate the ptr array within the arena if it needs to grow (because we can guarantee that a reallocation is going to require a `memcpy`, whereas in a normal libc heap it sometimes doesn’t), so might be good to always slightly oversize the ptr array for an average usecase. Or allow the caller to specify the expected array length, so the arena code doesn’t have to guess.